Geocoding using Google Maps API

ashmeet13 opened the pull request:
Adding scripts use to geocode a given sheet.

ashmeet13 committed:
Add scripts used to geocode Assam Travel data

ashmeet13 committed:
Add Geocoding readme

ashmeet13 committed:
Update readme, add comments in scripts

NirantK commented on geocoding/README.md:
```diff
@@ -0,0 +1,6 @@
+1. `pip install -U googlemaps`
+2. Change `xlsx_name = "ReportedTravel.xlsx"` with the name of your xlsx file or path to it.
+3. Enter your API key in the code.
+4. Run the script `python geocode.py`
```
Parameterize the script to take API Key as a terminal input using [Fire](https://github.com/google/python-fire) or [Click](https://click.palletsprojects.com/en/7.x/)? or from OS ENV variable?

ashmeet13 replied:
Parameterized variables as arguments using fire.

NirantK commented on geocoding/geocode.py:
```diff
@@ -0,0 +1,65 @@
+    print(f"{sheet_name} sheet has {len(addresses)} rows")
+    print()
+
+    for address in addresses:
```
`tqdm` or `fastprogress` this to get a sense of time needed?

ashmeet13 replied:
Added tqdm

NirantK commented on geocoding/geocode.py:
```diff
@@ -0,0 +1,65 @@
+import pandas as pd
```
isort and black?

ashmeet13 replied:
isort and black done .

NirantK commented on geocoding/geocode.py:
```diff
@@ -0,0 +1,65 @@
+import googlemaps
+import os
+
+if not os.path.isdir("Sheets"):
```
Parameterize both filename and API key perhaps?

ashmeet13 replied:
Done

NirantK commented on geocoding/geocode.py:
```diff
@@ -0,0 +1,65 @@
+import pandas as pd
+import googlemaps
+import os
```
Prefer `Pathlib` over `os` for cross-os consistency

ashmeet13 replied:
Removed `os`.
Added `Pathlib`

NirantK commented on geocoding/verify.py:
```diff
@@ -0,0 +1,20 @@
+'''
```
Excellent idea! Perhaps this should be called `tests`?

ashmeet13 replied:
Renamed to `tests`

ashmeet13 committed:
1. Change os to pathlib
2. Black and isort the scripts
3. Add fire support in geocode.py
4. Add script to merge all formed CSV's under one.

ashmeet13 commented:
@NirantK updated the branch with proposed changes. Requesting another review.

NirantK commented on geocoding/geocode.py:
```diff
@@ -1,65 +1,81 @@
+        df["Longitude"] = longs
+        df["Geocoded"] = geocoded
+
+        print()
+        print(f"Saving {sheet_name} with geocoded address")
+        print(f"Unable to geocode - {count}")
+        total_none_count += count
+        total_xlsx_count += total_count
+        print()
```
Blank print statements?

NirantK replied:
These should probably be logging statements going to `stdout` or a file or both? I believe the blank print statements were for debugging and are no longer necessary?

ashmeet13 replied:
Added logging rather than `print` statements.

ashmeet13 committed:
1. Add script to generate pincodes.
2. Add logging

NirantK commented on geocoding/generate_pincode.py:
```diff
@@ -0,0 +1,109 @@
+    nancount = 0
+    success = 0
+    index = 0
+    for lat, lng in tqdm(zip(lats, lngs)):
```
This is hard for me to read right now given there are 3+ flags. What can we do simplify?

One way plausible way: 
Can we separate these out different cases into neater functions -- even nested function will do.

ashmeet13 replied:
Cleaned the code and improved readability.
Added docstrings.

NirantK commented on geocoding/generate_pincode.py:
```diff
@@ -0,0 +1,109 @@
+            logging.warning(f"API returned no postal code value - {(lat,lng)}")
+
+    if len(set(pincode_verify)) > 1:
+        logging.info("There were failures when trying to reach the API.")
```
```suggestion
        logging.warn("There were failures when trying to reach the API.")
```

Maybe this should be a warning?

ashmeet13 replied:
Changed to warning

NirantK commented on geocoding/generate_pincode.py:
```diff
@@ -0,0 +1,109 @@
+        )
+        df["pincode"] = postal_codes
+        df['is_assam'] = is_assam
+        df.to_csv("AssamNewAddressPincode.csv")
```
I am not a huge fan of hardcoding filenames. If we want a default, maybe we can take this output filename as a function input? This would allow you to set it via terminal via `Fire`. 

Something on these lines:

```python
def reverse_geocode(csv_sheet, api_key, output_filename="Assam_Geocoded_wPincode.csv"):
```

ashmeet13 replied:
Parameterized the filename.

NirantK commented on geocoding/geocode.py:
```diff
@@ -0,0 +1,125 @@
+logging.info("Logger Setup Complete")
+
+
+def make_dir():
+    """Responsible for creating the folder in which the CSV's will
+    be stored in.
+    """
+    sheets_dir = pathlib.Path.cwd() / "Sheets"
+
+    if not sheets_dir.is_dir():
+        sheets_dir.mkdir()
```
You can do this inline in 2 lines:

```suggestion
from pathlib import Path
sheets_dir:Path = Path.cwd() / "Sheets"
sheets_dir.mkdir(parents=True, exist_ok=True)
```

Or even one line:
```python
(Path.cwd() / "Sheets").mkdir(parents=True, exist_ok=True)
```

ashmeet13 replied:
Removed the overly elaborate function. Replaced with - 
`(Path.cwd() / "Sheets").mkdir(parents=True, exist_ok=True)`

ashmeet13 committed:
Clean code in generate_pincode.py and small change in geocode.py

NirantK commented on geocoding/generate_pincode.py:
```diff
@@ -18,91 +18,153 @@
+        successful_api_hit = test[index]
+        index += 1
+
+        if not successful_api_hit:
```
What are we trying to do here? I don't follow.

ashmeet13 replied:
This section of code a few pieces of code needed to be removed.
Made the changes and added comments to explain the flow.

NirantK commented on geocoding/generate_pincode.py:
```diff
@@ -18,91 +18,153 @@
             pincode_verify.append(False)
             postal_codes.append(None)
             is_assam.append(None)
+            logging.info("Here??")
```
Do you think we can probably have something more descriptive -- other than `Here??` ?

ashmeet13 replied:
My bad. Removed the logging.

NirantK commented on geocoding/geocode.py:
```diff
@@ -112,14 +102,14 @@ def geocode(data_file, api_key, use_column="Address"):
         use_column (str, optional): The column that is to be used as input to API while geocoding. Defaults to "Address".
     """
-    make_dir()
+    (pathlib.Path.cwd() / "Sheets").mkdir(parents=True, exist_ok=True)
```
```suggestion
    (pathlib.Path.cwd() / "Sheets").mkdir(parents=True, exist_ok=True) # make directory "Sheets" if it does not exist
```

ashmeet13 replied:
Added the change.

NirantK commented on geocoding/geocode.py:
```diff
@@ -112,14 +102,14 @@ def geocode(data_file, api_key, use_column="Address"):
 
     if data_file.endswith(".json"):
         geocode_json(data_file, gmaps, use_column)
     else:
```
```suggestion
    elif data_file.endswith(".xlsx") or data_file.endswith(".xls"):
```
Explicit is better than implicit.

ashmeet13 replied:
Added the change.

NirantK commented on geocoding/clean.py:
```diff
@@ -0,0 +1,30 @@
+    temp = pd.ExcelFile(excel_sheet)
+    sheet_names = temp.sheet_names
+
+    maindf = pd.read_csv(f"Sheets/{sheet_names[0]}.csv")
```
```suggestion
    main_df = pd.read_csv(f"Sheets/{sheet_names[0]}.csv")
```
1. Always use snake case variable naming please? Or use a linter like autoPEP8. 
2. Can we use more specific variables like like `xl_pd_buffer` (instead of `temp`) and `sheet_df` (instead of `main_df`) so that they are clearer?

ashmeet13 replied:
Got it. Renamed the variables.

NirantK commented on geocoding/clean.py:
```diff
@@ -0,0 +1,30 @@
+    maindf = maindf.drop("Geocoded", axis=1)
+
+    total = len(maindf)
+    for sheet_name in sheet_names[1:]:
```
What is the reason for starting from 1 and not 0?

ashmeet13 replied:
Refactored this file and renamed to a more appropriate name. Added comments to explain.

NirantK commented on geocoding/generate_pincode.py:
```diff
@@ -0,0 +1,171 @@
+import pandas as pd
+from tqdm import tqdm
+
+logging.basicConfig(
+    filename="pincoding.log",
+    filemode="a",
+    format="%(asctime)s,%(msecs)d %(name)s %(levelname)s %(message)s",
+    datefmt="%H:%M:%S",
+    level=logging.INFO,
+)
+
+logging.info("Logger Setup Complete")
```
Maybe move the logging config to a separate file and use that?

Maybe consider using Logaru? https://github.com/Delgan/loguru
Or Structlog -- which is what we use at @verloop under the hood behind Verloop-CTL?

ashmeet13 replied:
Shifted the config to another file.
Switched to loguru.

NirantK commented on geocoding/generate_pincode.py:
```diff
@@ -0,0 +1,171 @@
+    Returns:
+        dict: Response recieved from API
+    """
+    successful_api_hit = True
```
`successful_api_hit` can be more precise? Is this 200 OK? Or that you got valid data back? It'd help if the variable name distinguished it.

ashmeet13 replied:
Replaced the name `successful_api_hit` with `valid_data_received`

NirantK commented on geocoding/generate_pincode.py:
```diff
@@ -0,0 +1,171 @@
+        lng (float): Longitude
+        gmaps (Object): gmaps object with declared API key
+    
+    Returns:
```
Outdated docstring on `Returns`

ashmeet13 replied:
Fixed the docstring

NirantK commented on geocoding/generate_pincode.py:
```diff
@@ -0,0 +1,171 @@
+        output_file (str, optional): Name of the new CSV to be stored as. Defaults to 'AssamNewAddressPincode'.
+    """
+    gmaps = googlemaps.Client(key=api_key)
+    csv_sheet = pathlib.Path.cwd() / csv_sheet
```
```suggestion
    csv_sheet = pathlib.Path.cwd() / csv_sheet
```

What can we do to remove the current directory compulsion? Maybe just take the complete path from function input? And then we can resolve it? And we won't have to prepend the `cwd`?

ashmeet13 replied:
Removed the `cwd` compulsion. Works with `Path.resolve()`.

ashmeet13 committed:
PR suggested changes pushed

NirantK commented on geocoding/geocode.py:
```diff
@@ -71,9 +71,9 @@ def geocode_json(data_file, gmaps, use_column):
-    xlsx_name = pathlib.Path.cwd() / data_file
-    temp = pd.ExcelFile(xlsx_name)
-    sheet_names = temp.sheet_names
+    xlsx_name = data_file.resolve()
```
Yay!

ashmeet13 committed:
Better variable name

ashmeet13 committed:
Typo fixed

ashmeet13 committed:
Abstract logger config in different file. Switch to loguru
